design/ceph/ceph-orchestrator-osd-add-remove.md

```diff
@@ -0,0 +1,266 @@
+of failure. Rook may report successful statuses back if needed by the Orchestrator Module; however,
+limiting the information reported back will keep the `CephCluster` resource's `status` field smaller
+and less unwieldy for users.
+
+```yaml
+# first draft suggestions for status mechanism
+apiVersion: ceph.rook.io/v1
+kind: CephCluster
+status:
+  driveGroupOSDCreations: # proposed new
+    - node: < string, node name > # proposed new
+      driveGroups: # proposed new
+        - spec: < string, JSON or YAML blob from storage spec >
+          success: < bool, true if successful, false if not >
+          message: < string, message from results of last spec application >
+```
+
+A failure status for a Drive Group on a node should be cleared when the Drive Group is run
+successfully by `ceph-volume`.
+
+The failure status of the last-applied `ceph-volume` command with the Drive Group spec is
+reported in `status`, and a corresponding `message` is also reported. The message must be filled in
+if the status is `false` (unsuccessful).
+
+Even though successful Drive Group applications are not reported `success` is still included in the
+status so that the failure is clear to human eyes.
+
+### Removing OSDs
+When removing an OSD (or OSDs) via the Rook Orchestrator Module, the Ceph mgr running in Kubernetes
+will also perform whatever Ceph operations are necessary to remove the OSD as would be expected by
+the user. This will replace the manual, imperative steps currently required by Rook-Ceph users to
+remove OSDs from the storage cluster documented
+[here for Rook v1.2](https://rook.io/docs/rook/v1.2/ceph-osd-mgmt.html#remove-an-osd).
+
+There is currently a proposal for a shared Ceph Mgr Orchestrator Module implementing common OSD
+removal steps [[link]](https://github.com/ceph/ceph/pull/32677) usable by any other Ceph Mgr
+Orchestrator module. The OSD removal mechanism should use this where possible.
+
+Rook will not automatically remove Kubernetes Deployments for OSDs unless
+`removeOSDsIfOutAndSafeToRemove: true`. This should be documented for users, and users should be
+advised to set this value to `true`.
+
+The `CephCluster` CRD allows users to add disks to nodes in several places including disks on all
+nodes in the Kubernetes cluster, to a specific node, or via disk pattern matching (in two different
+ways). To avoid the complexity of having to manage the logic for determining how to remove a node
+from the `CephCluster` resource, OSD removal will **not** alter the `CephCluster` resource's
+`storage` specification with the exception of Drive Groups when needed. While modifying these items
+might be convenient for users, there are a lot of opportunities in development to miss corner cases
+or introduce logic bugs, which we should avoid.
+
+Users should be advised in documentation that if they plan to use the Ceph CLI or Dashboard to
+manage OSD addition in the Rook-Ceph cluster, it is recommended to set `useAllDevices: false`, leave
```

Comment:
Leave this in the hands of the user that reads the documentation seems to me optimistic.  
This is going very hard/impossible to solve because the root problem is that we are using and "abstract" selection of devices by different methods. If we would have in the cluster CRD one item declared by each OSD most of the problems we have would disappear.

Reply:
As @travisn mentioned, we could design the mgr module to be "opinionated" and force `useAllDevices: false`, set disk pattern matching fields empty, and remove disks specified by node when users use the orchestrator interface(s).

Reply:
This means that in documentation we should include something like:
"If in any moment you wish to add/remove OSD's, then you have to  set "useAllDevices: false" in the Rook cluster CRD. Once it has been set to False this setting cannot be reverted, because the the deleted OSD's will  be created again"
Or... in the case that we want to set this flag automatically... in what cases we should set it to "False", how do we detect this cases?